lib/gui/components/drive-selector/controllers/drive-selector.js

```diff
@@ -170,4 +171,81 @@ module.exports = function(
     });
   };
 
+  /**
+   * @summary Memoize ImmutableJS list reference
+   * @function
+   * @private
+   *
+   * @description
+   * This workaround is needed to avoid AngularJS from getting
+   * caught in an infinite digest loop when using `ngRepeat`
+   * over a function that returns a mutable version of an
+   * ImmutableJS object.
+   *
+   * The problem is that every time you call `myImmutableObject.toJS()`
+   * you will get a new object, whose reference is different from
+   * the one you previously got, even if the data is exactly the same.
+   *
+   * @param {Function} func - function that returns an ImmutableJS list
+   * @param {Function} compare - function used to compare the old and new lists
+   * @returns {Function} memoized function
+   *
+   * @example
+   * const getList = () => {
+   *   return Store.getState().toJS().myList;
+   * };
+   *
+   * const memoizedFunction = memoizeImmutableListReference(getList);
+   */
+  const memoizeImmutableListReference = (func, compare = _.isEqual) => {
+    let previous = [];
+
+    return (...restArgs) => {
+      const list = Reflect.apply(func, this, restArgs);
+
+      if (!_.isEqualWith(list, previous, compare)) {
+        previous = list;
+      }
+
+      return previous;
+    };
+  };
+
+  /**
+   * @summary Get the list of drives and their compatibility status objects
+   * @function
+   * @public
+   *
+   * @description
+   * For each drive, return its compatibility status with the seleted image,
+   * containing the status type (ERROR, WARNING), and accompanying
+   * status message.
+   *
+   * @returns {Object[]} list of objects containing drive and statuses
+   *
+   * @example
+   * const metadrives = DriveSelectorController.getDrivesAndStatuses();
+   *
+   * for ({ drive, statuses } of metadrives) {
+   *   for ({ type, message } of statuses) {
+   *     // do something
+   *   }
+   * }
+   */
+  this.getDrivesAndStatuses = memoizeImmutableListReference(() => {
+    const statusList = _.map(this.drives.getDrives(), (drive) => {
+      return {
+        drive,
+        statuses: this.constraints.getDriveImageCompatibilityStatuses(drive, this.state.getImage())
+      };
+    });
+
+    return statusList;
+  }, (newList, oldList) => {
+    return _.isEqual(
+      utils.removeAngularHashes(newList),
```

Comment:
OK, I had great success with this! Here's a diff with my local changes:

```diff
diff --git a/lib/gui/components/drive-selector/controllers/drive-selector.js b/lib/gui/components/drive-selector/controllers/drive-selector.js
index df8fc8d..1579ed0 100644
--- a/lib/gui/components/drive-selector/controllers/drive-selector.js
+++ b/lib/gui/components/drive-selector/controllers/drive-selector.js
@@ -16,6 +16,7 @@
 
 'use strict';
 
+const angular = require('angular');
 const _ = require('lodash');
 const messages = require('../../../../shared/messages');
 const constraints = require('../../../../shared/drive-constraints');
@@ -197,13 +198,13 @@ module.exports = function(
    *
    * const memoizedFunction = memoizeImmutableListReference(getList);
    */
-  const memoizeImmutableListReference = (func, compare = _.isEqual) => {
+  const memoizeImmutableListReference = (func) => {
     let previous = [];
 
     return (...restArgs) => {
       const list = Reflect.apply(func, this, restArgs);
 
-      if (!_.isEqualWith(list, previous, compare)) {
+      if (!angular.equals(list, previous)) {
         previous = list;
       }
 
@@ -211,41 +212,12 @@ module.exports = function(
     };
   };
 
-  /**
-   * @summary Get the list of drives and their compatibility status objects
-   * @function
-   * @public
-   *
-   * @description
-   * For each drive, return its compatibility status with the seleted image,
-   * containing the status type (ERROR, WARNING), and accompanying
-   * status message.
-   *
-   * @returns {Object[]} list of objects containing drive and statuses
-   *
-   * @example
-   * const metadrives = DriveSelectorController.getDrivesAndStatuses();
-   *
-   * for ({ drive, statuses } of metadrives) {
-   *   for ({ type, message } of statuses) {
-   *     // do something
-   *   }
-   * }
-   */
-  this.getDrivesAndStatuses = memoizeImmutableListReference(() => {
-    const statusList = _.map(this.drives.getDrives(), (drive) => {
-      return {
-        drive,
-        statuses: this.constraints.getDriveImageCompatibilityStatuses(drive, this.state.getImage())
-      };
-    });
+  this.getDrives = memoizeImmutableListReference(() => {
+    return this.drives.getDrives();
+  });
 
-    return statusList;
-  }, (newList, oldList) => {
-    return _.isEqual(
-      utils.removeAngularHashes(newList),
-      utils.removeAngularHashes(oldList)
-    );
+  this.getDriveStatuses = memoizeImmutableListReference((drive) => {
+    return this.constraints.getDriveImageCompatibilityStatuses(drive, this.state.getImage());
   });
 
 };
diff --git a/lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html b/lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html
index 690782a..3525ea6 100644
--- a/lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html
+++ b/lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html
@@ -5,17 +5,17 @@
 
 <div class="modal-body">
   <ul class="list-group">
-    <li class="list-group-item" ng-repeat="meta in modal.getDrivesAndStatuses() track by meta.drive.device"
-      ng-disabled="!modal.constraints.isDriveValid(meta.drive, modal.state.getImage())"
-      ng-dblclick="modal.selectDriveAndClose(meta.drive)"
-      ng-click="modal.toggleDrive(meta.drive)">
+    <li class="list-group-item" ng-repeat="drive in modal.getDrives() track by drive.device"
+      ng-disabled="!modal.constraints.isDriveValid(drive, modal.state.getImage())"
+      ng-dblclick="modal.selectDriveAndClose(drive)"
+      ng-click="modal.toggleDrive(drive)">
         <div>
-          <h4 class="list-group-item-heading">{{ meta.drive.description }} - {{ meta.drive.size | gigabyte | number:1 }} GB</h4>
-          <p class="list-group-item-text">{{ meta.drive.name }}</p>
+          <h4 class="list-group-item-heading">{{ drive.description }} - {{ drive.size | gigabyte | number:1 }} GB</h4>
+          <p class="list-group-item-text">{{ drive.name }}</p>
 
           <footer class="list-group-item-footer">
 
-            <span class="label" ng-repeat="status in meta.statuses"
+            <span class="label" ng-repeat="status in modal.getDriveStatuses(drive)"
               ng-class="{
                 'label-warning': status.type === modal.constraints.COMPATIBILITY_STATUS_TYPES.WARNING,
                 'label-danger': status.type === modal.constraints.COMPATIBILITY_STATUS_TYPES.ERROR
@@ -26,8 +26,8 @@
           </footer>
         </div>
         <span class="tick tick--success"
-          ng-show="modal.constraints.isDriveValid(meta.drive, modal.state.getImage())"
-          ng-disabled="!modal.state.isCurrentDrive(meta.drive.device)"></span>
+          ng-show="modal.constraints.isDriveValid(drive, modal.state.getImage())"
+          ng-disabled="!modal.state.isCurrentDrive(drive.device)"></span>
     </li>
     <li class="list-group-item"
       ng-show="!modal.drives.hasAvailableDrives()">
diff --git a/lib/shared/utils.js b/lib/shared/utils.js
index 337c7f9..c9e98e3 100644
--- a/lib/shared/utils.js
+++ b/lib/shared/utils.js
@@ -76,32 +76,3 @@ exports.percentageToFloat = (percentage) => {
 
   return percentage / PERCENTAGE_MAXIMUM;
 };
-
-/**
- * @summary Traverse an object and remove Angular hashes
- * @function
- * @public
- *
- * @param {Object} object - object
- * @returns {Object} object
- *
- * @example
- * const angularObjA = { $$hashKey: 'abc' };
- * const angularObjB = { $$hashKey: '123' };
- *
- * const areObjectsEqual = _.isEqual(
- *   utils.removeAngularHashes(angularObjA),
- *   utils.removeAngularHashes(angularObjB)
- * );
- */
-exports.removeAngularHashes = (object) => {
-  if (_.isArray(object)) {
-    return _.map(object, exports.removeAngularHashes);
-  }
-  if (_.isObject(object)) {
-    const omitted = _.omit(object, [ '$$hashKey' ]);
-    return _.mapValues(omitted, exports.removeAngularHashes);
-  }
-
-  return object;
-};
diff --git a/tests/shared/utils.spec.js b/tests/shared/utils.spec.js
index cea0b1d..fd604ec 100644
--- a/tests/shared/utils.spec.js
+++ b/tests/shared/utils.spec.js
@@ -131,74 +131,4 @@ describe('Shared: Utils', function() {
 
   });
 
-  describe('.removeAngularHashes()', function() {
-
-    it('should remove direct hashes', function() {
-      m.chai.expect(utils.removeAngularHashes({
-        attrA: 1,
-        $$hashKey: '123'
-      })).to.deep.equal({
-        attrA: 1
-      });
-    });
-
-    it('should remove hashes from objects in list', function() {
-      m.chai.expect(utils.removeAngularHashes([
-        {
-          attrA: 1,
-          $$hashKey: '123'
-        },
-        {
-          attrB: 2,
-          $$hashKey: 'abc'
-        }
-      ])).to.deep.equal([
-        {
-          attrA: 1
-        },
-        {
-          attrB: 2
-        }
-      ]);
-    });
-
-    it('should remove hashes from objects in objects', function() {
-      m.chai.expect(utils.removeAngularHashes({
-        attrA: {
-          nestedA: 1,
-          $$hashKey: '123'
-        },
-        attrB: {
-          nestedB: 2,
-          $$hashKey: 'abc'
-        },
-        $$hashKey: 'xyz'
-      })).to.deep.equal({
-        attrA: {
-          nestedA: 1
-        },
-        attrB: {
-          nestedB: 2
-        }
-      });
-    });
-
-    it('should return identity when no hashes', function() {
-      m.chai.expect(utils.removeAngularHashes({
-        attrA: 1
-      })).to.deep.equal({
-        attrA: 1
-      });
-    });
-
-    it('should return identity when not an object or list', function() {
-      m.chai.expect(utils.removeAngularHashes(1)).to.deep.equal(1);
-    });
-
-    it('should return identity when null', function() {
-      m.chai.expect(utils.removeAngularHashes(null)).to.deep.equal(null);
-    });
-
-  });
-
 });
```

This is what I did:

- Always use `angular.equals` to compare objects in the `memoizeImmutableListReference` function, since this function already deals with angular hashes correctly
- Thus, we can get rid of the whole `utils.removeAngularHashes` function
- Create two thin functions that simply wrap `drives.getDrives` and `constraints.getDriveImageCompatibilityStatuses` with `memoizeImmutableListReference`
- Use these functions on the template directly, without creating a temporary data structure to mix both

What do you think?